### PR TITLE
Merge development into marketplace

### PR DESCRIPTION
**Pull request by spencern**
Merges latest `development` into `marketplace`
Resolves all conflicts (a lot!)

### CODE REVIEW

File: imports/plugins/core/ui/client/components/tags/tags.jsx [hunk 1]
```diff
@@ -102,13 +103,13 @@ class Tags extends Component {
...
               onMove={this.props.onMoveTag}
               draggable={this.props.draggable}
-              onTagClick={this.handleTagClick}
               onTagInputBlur={this.handleTagSave}
               onTagMouseOut={this.handleTagMouseOut}
               onTagMouseOver={this.handleTagMouseOver}
               onTagRemove={this.handleTagRemove}
               onTagSave={this.handleTagSave}
               onTagUpdate={this.handleTagUpdate}
+              onTagClick={this.handleTagClick}
```
Comment (spencern) [inline imports/plugins/core/ui/client/components/tags/tags.jsx]: Do these need to be in alphabetical order?

File: imports/plugins/included/product-variant/client/templates/products/productDetail/variants/variantForm/variantForm.js [hunk 2]
```diff
@@ -6,7 +6,7 @@ import { Reaction, i18next } from "/client/api";
 import { ReactionProduct } from "/lib/api";
 import { applyProductRevision } from "/lib/api/products";
 import { Products } from "/lib/collections";
-import VariantFormContainer from "../../../../../containers/variantFormContainer";
+import VariantFormContainer from "/imports/plugins/included/product-variant/containers/variantFormContainer";
```
Comment (spencern) [inline imports/plugins/included/product-variant/client/templates/products/productDetail/variants/variantForm/variantForm.js]: 👌

File: client/modules/accounts/templates/profile/profile.js [hunk 3]
```diff
@@ -67,13 +68,5 @@ Template.accountProfile.helpers({
       return "addressBookGrid";
     }
     return "addressBookAdd";
-  },
-
-  isMarketplaceGuest: function () {
```
Comment (spencern) [inline client/modules/accounts/templates/profile/profile.js]: reverted

**Comment by spencern** [comment]
Sticking the WIP tag back on this. There are some elements of marketplace that are broken by this PR. Still sorting through it.

**Comment by spencern** [comment]
This should be ready to go.

**Comment by spencern** [comment]
So this is actually still not ready to go, something got screwed up with permissions for merchants and now can't update shop details for owned shop. Still working through it. :(

**Comment by spencern** [comment]
I _swear_ this worked before, but I’m not sure what got changed and when I go back to older commits, the same issue is present.
The apparent problem is that the Security.permit / Security.deny collection permission security methods are running before a users Id has been loaded.

The fix here is to create a new security method `ifHasRoleForActiveShop` that moves checking `Reaction.getShopId()` to the security method instead of on the `permit` or `deny` method.

@zenweasel would you check this out when you get a chance. I _think_ it's ready to merge to marketplace now and would you review the changes in `/server/security/collections.js` specifically.

After this merges I think we should be able to get the team to start working off of the marketplace branch instead of `development` for most tasks.

**Comment by zenweasel** [comment]
@spencern What exactly did you want me to review? Ddi you want me to test it?

**Comment by spencern** [comment]
@zenweasel Yeah, I should have put "to test" steps up.

I was specifically hoping you could look at the new security method I added to make the active shop admin permissions work correctly for client side database ops.

https://github.com/reactioncommerce/reaction/blob/spencer-merge-dev-marketplace/server/security/collections.js#L43-L53

I'm pretty sure it's fine, but seems like the type of code that should be reviewed.

**Comment by spencern** [comment]
Additionally,

To test:
1. Start server
2. Create a new account
3. Click "Become a Seller" in the users profile
4. Change the name or other shop details of the new shop that gets created.

File: imports/plugins/included/shippo/client/settings/shippoTableColumn.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import React, { Component } from "react";
+import PropTypes from "prop-types";
+import { Icon } from "/imports/plugins/core/ui/client/components";
+
+class EmailTableColumn extends Component {
```
Comment (zenweasel) [inline imports/plugins/included/shippo/client/settings/shippoTableColumn.js]: This is probably in `development` but why is this called `EmailTableColumn`?

**Comment by zenweasel** [comment]
If I create an account and then click "Become a seller" I get a bunch of console errors:

```
TypeError: Cannot read property 'brandAssets' of undefined
    at Brand.getLogo (brand.js:21)
    at Brand.render (brand.js:30)
    at modules.js?hash=1e7eb5b…:107993
    at measureLifeCyclePerf (modules.js?hash=1e7eb5b…:107272)
    at ReactCompositeComponentWrapper._renderValidatedComponentWithoutOwnerOrContext (modules.js?hash=1e7eb5b…:107992)
    at ReactCompositeComponentWrapper._renderValidatedComponent (modules.js?hash=1e7eb5b…:108019)
    at ReactCompositeComponentWrapper._updateRenderedComponent (modules.js?hash=1e7eb5b…:107943)
    at ReactCompositeComponentWrapper._performComponentUpdate (modules.js?hash=1e7eb5b…:107921)
    at ReactCompositeComponentWrapper.updateComponent (modules.js?hash=1e7eb5b…:107842)
    at ReactCompositeComponentWrapper.receiveComponent (modules.js?hash=1e7eb5b…:107744)
```

It repeats with can't find `locales`, `language` of `undefined`

**Comment by spencern** [comment]
@zenweasel I'd seen those errors before but couldn't reproduce them. I'll see if I can address those now

**Comment by zenweasel** [comment]
Testing this again with a full reset I am not seeing it and can't reproduce it either. I had done a full reset before as well so I don't know what the deal it.

File: package.json [hunk 5]
```diff
@@ -138,24 +141,35 @@
...
   "jest": {
     "moduleNameMapper": {
-      "^\/lib(.*)": "<rootDir>/lib/$1",
-      "^\/imports\/plugins(.*)": "<rootDir>/imports/plugins/$1",
-      "^\/client\/api(.*)": "<rootDir>/imports/test-utils/__mocks__/client/api$1",
+      "^/lib(.*)": "<rootDir>/lib/$1",
+      "^/imports/plugins(.*)": "<rootDir>/imports/plugins/$1",
+      "^/client/api(.*)": "<rootDir>/imports/test-utils/__mocks__/client/api$1",
       "^meteor/aldeed:simple-schema": "<rootDir>/imports/test-utils/__mocks__/meteor/aldeed-simple-schema",
+      "^meteor/tmeasday:publish-counts": "<rootDir>/imports/test-utils/__mocks__/meteor/aldeed-simple-schema",
```
Comment (zenweasel) [inline package.json]: I think this is wrong.

Comment (spencern) [inline package.json]: Nice catch.

Comment (zenweasel) [inline package.json]: Looks like this is wrong in `development` too

Comment (spencern) [inline package.json]: Actually, identical to dev.. https://github.com/reactioncommerce/reaction/blob/development/package.json#L144-L149

Comment (spencern) [inline package.json]: Still feels wrong though.

File: server/api/core/core.js [hunk 6]
```diff
@@ -85,7 +89,7 @@ export default {
   },
   /**
    * registerTemplate
-   * registers Templates into the Templates Collection
+   * registers Templates into the Tempaltes Collection
```
Comment (zenweasel) [inline server/api/core/core.js]: typo

File: server/api/core/core.js [hunk 7]
```diff
@@ -311,60 +303,52 @@ export default {
...
-    Logger.info(`Adding defaultSellerRoles permissions for ${roles}`);
+  insertPackagesForShop(shopId) {
+    const layouts = [];
+    if (!shopId) {
+      return [];
+    }
+    const packages = this.Packages;
+    // for each shop, we're loading packages in a unique registry
+    // Object.keys(pkgConfigs).forEach((pkgName) => {
+    for (const packageName in packages) {
```
Comment (zenweasel) [inline server/api/core/core.js]: if `packages` is an array can you just use `for..of` to iterate?

File: server/publications/collections/product.js [hunk 8]
```diff
@@ -72,103 +50,75 @@ export function findProductMedia(publicationInstance, productIds) {
...
+      {
+        ancestors: {
+          $in: [_id]
+        }
+      }
+    ]
+  };
+
+  // Authorized content curators fo the shop get special publication of the product
+  // all all relevant revisions all is one package
```
Comment (zenweasel) [inline server/publications/collections/product.js]: Typo in comment

File: server/api/core/core.js (hunk 7 above)
Comment (spencern) [inline server/api/core/core.js]: probably could, seems like that would be better. Could lose the guard then I think.

File: server/publications/collections/product.js [hunk 9]
```diff
@@ -271,7 +216,7 @@ Meteor.publish("Product", function (productId) {
     ];
   }
 
-  // Everyone else gets the standard, visible products and variants
+  // Everyone else gets the standard, visibile products and variants
```
Comment (zenweasel) [inline server/publications/collections/product.js]: Changed from correctly spelled to incorrect

File: server/api/core/core.js (hunk 7 above)
Comment (spencern) [inline server/api/core/core.js]: actually, packages is an object where the keys are the packageName, not an array.

File: server/publications/collections/product.js (hunk 9 above)
Comment (spencern) [inline server/publications/collections/product.js]: hah, wonder how that snuck in.

**Comment by zenweasel** [comment]
I expect that everything I noted is probably also on `development` so this looks good to me. I checked the code you noted @spencern and it seems fine to me as far as my understanding of what it's supposed to do.

**Pull request merged**